components/style/restyle_hints.rs

```diff
@@ -383,38 +405,79 @@ impl DependencySet {
     }
 
     pub fn clear(&mut self) {
-        self.deps.clear();
+        self.common_deps.clear();
+        self.attr_deps.clear();
+        self.state_deps.clear();
     }
 }
 
 impl DependencySet {
+    fn test_dependency<E>(&self,
+                          dep: &Dependency,
+                          element: &E,
+                          snapshot: &ElementWrapper<E>,
+                          state_changes: &ElementState,
+                          attrs_changed: bool,
+                          hint: &mut RestyleHint)
+        where E: ElementExt
+    {
+        if state_changes.intersects(dep.sensitivities.states) || (attrs_changed && dep.sensitivities.attrs) {
+            let matched_then =
+                matches_complex_selector(&dep.selector, snapshot, None, &mut StyleRelations::empty());
+            let matches_now =
+                matches_complex_selector(&dep.selector, element, None, &mut StyleRelations::empty());
+            if matched_then != matches_now {
+                hint.insert(combinator_to_restyle_hint(dep.combinator));
+            }
+        }
+    }
+
     pub fn compute_hint<E>(&self, el: &E,
                            snapshot: &E::Snapshot,
                            current_state: ElementState)
                            -> RestyleHint
-    where E: ElementExt + Clone
+        where E: ElementExt + Clone
     {
         debug!("About to calculate restyle hint for element. Deps: {}",
-               self.deps.len());
+               self.len());
 
         let state_changes = snapshot.state().map_or_else(ElementState::empty, |old_state| current_state ^ old_state);
         let attrs_changed = snapshot.has_attrs();
+
+        if state_changes.is_empty() && !attrs_changed {
+            return RestyleHint::empty();
+        }
+
         let mut hint = RestyleHint::empty();
-        for dep in &self.deps {
-            if state_changes.intersects(dep.sensitivities.states) || (attrs_changed && dep.sensitivities.attrs) {
-                let old_el: ElementWrapper<E> = ElementWrapper::new_with_snapshot(el.clone(), snapshot);
-                let matched_then =
-                    matches_complex_selector(&*dep.selector, &old_el, None, &mut StyleRelations::empty());
-                let matches_now =
-                    matches_complex_selector(&*dep.selector, el, None, &mut StyleRelations::empty());
-                if matched_then != matches_now {
-                    hint.insert(combinator_to_restyle_hint(dep.combinator));
-                    if hint.is_all() {
-                        break
-                    }
+        let snapshot = ElementWrapper::new_with_snapshot(el.clone(), snapshot);
+        for dep in &self.common_deps {
+            self.test_dependency(dep, el, &snapshot, &state_changes,
+                                 attrs_changed, &mut hint);
+            if hint.is_all() {
+                return hint;
+            }
+        }
+
+        if !state_changes.is_empty() {
```

Comment:
Rather than doing this in three places here, I think this should be abstracted into a method on DependencySet.